salt/_states/metalk8s_kubernetes.py

```diff
@@ -1572,3 +1576,205 @@ def node_uncordoned(name, **kwargs):
         the name of the node
     '''
     return _node_unschedulable(name, False, **kwargs)
+
+
+def _extract_custom_resource_data(name, namespace, body, source, **kwargs):
+    if body and source:
+        raise salt.exceptions.InvalidConfigError(
+            "'source' cannot be used in combination with 'body'"
+        )
+
+    if source:
+        try:
+            with open(source, 'r') as source_file:
+                body = yaml.safe_load(source_file)
+        except IOError as exc:
+            raise salt.exceptions.InvalidConfigError(
+                "no such '{0}' source file".format(source)
+            )
+
+    # TODO: handle kindlist
+    try:
+        kind = body['kind']
+    except KeyError:
+        raise salt.exceptions.InvalidConfigError(
+            "object 'kind' is missing in object definition"
+        )
+
+    try:
+        group, version = body.get('apiVersion', '').split('/')
+    except ValueError:
+        raise salt.exceptions.InvalidConfigError(
+            "malformed or undefined 'apiVersion' in object definition"
+        )
+
+    plural = __salt__['metalk8s_kubernetes.get_crd_plural_from_kind'](
+        kind, **kwargs)
+    if plural is None:
+        raise salt.exceptions.InvalidConfigError(
+            "'{0}' object kind not found in API".format(kind)
```

Comment:
So, this error message isn't really explaining the failed check above... In any case, is it really mandatory for CRDs to define a `plural`? I'm not sure it's something that the Operator SDK generates, for example.

Reply:
I'm not sure if it's mandatory, maybe if you create a CRD without plural it'll fallback on `kind` value.
I'll check that.

Reply:
Looks like it's not possible to create a CRD without `spec.names.plural`:
```
# kubectl apply -f 0prometheus-operator-0servicemonitorCustomResourceDefinition.yaml
error: error validating "0prometheus-operator-0servicemonitorCustomResourceDefinition.yaml": error validating data: ValidationError(CustomResourceDefinition.spec.names): missing required field "plural" in io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1beta1.CustomResourceDefinitionNames; if you choose to ignore these errors, turn validation off with --validate=false
```